kafka/consumer.go

```diff
@@ -102,6 +109,15 @@ func NewConsumer(cfg ConsumerConfig) (*Consumer, error) {
 		kgo.ConsumeTopics(cfg.Topics...),
 		kgo.WithLogger(kzap.New(cfg.Logger)),
 	}
+	switch cfg.Delivery {
+	case apmqueue.AtLeastOnceDeliveryType:
+		// franz-go defaults to at least once processing, using auto committing.
+	case apmqueue.AtMostOnceDeliveryType:
+		// GreedyAutoCommit opts in to committing everything that has been polled
+		// when autocommitting (the dirty offsets), rather than committing what
+		// has previously been polled.
+		opts = append(opts, kgo.GreedyAutoCommit())
```

Comment:
Auto-commit seems to be done asynchronously, in a background goroutine: https://github.com/twmb/franz-go/blob/3322f311fd833129f91b4ea76581150274bdb457/pkg/kgo/consumer_group.go#L2009

So I don't think setting GreedyAutoCommit is enough to guarantee at most once delivery. I think we would need to disable autocommit, and manually commit before consuming the fetches.

Reply:
ah, you're right. I'll switch to manual commit then.

Reply:
I've made the required changes and added blocking rebalances during record a poll and until `client.AllowRebalance()` is called (to avoid committing offsets for partitions lost).